### PR TITLE
Fall back to the IP address that already successfully connected to

### PR DESCRIPTION
**Pull request by dHannasch**
This is built atop https://github.com/ray-project/ray/pull/12026, which is built atop https://github.com/ray-project/ray/pull/12022, so this *looks* very large and complicated at the moment but there's really only one simple change happening here. Most of the code is really just from https://github.com/ray-project/ray/pull/12022. (https://github.com/ray-project/ray/pull/12022 is conceptually simple but has very fiddly code; this changes only a tiny but of code, but the considerations about the change in behavior can be complex.)

EDIT I merged https://github.com/ray-project/ray/pull/13367 into this, too. That's built atop https://github.com/ray-project/ray/pull/12976.

https://github.com/ray-project/ray/pull/12026 emits a warning. But there isn't anything the user can actually do with the information, except "go complain to the Ray developers". This attempts to automatically fall back and recover. (This still might fail, in which case the warning will still be useful.)

(This write-up also looks very long and complicated, but seriously, the actual change to behavior is very simple. This is just a lot of background.)

## Why are these changes needed?

Background: currently, Ray stashes the addresses of the Redis shards under the special key `RedisShards`.
https://github.com/ray-project/ray/blob/master/python/ray/_private/services.py#L888
```
shard_address = address(node_ip_address, redis_shard_port)
rpush("RedisShards", shard_address)
```
When the head node starts, it rpushes the IP address and port for each shard. (In theory. Actually it rpushes its own IP address, but we'll get to that.)
https://github.com/ray-project/ray/blob/master/src/ray/gcs/redis_client.cc#L62
`redisCommand(context, "LRANGE RedisShards 0 -1")`
Each non-head node first connects to Redis at the head node, then retrieves the RedisShards from Redis and attempts to connect to each of them.
https://github.com/ray-project/ray/blob/master/src/ray/gcs/redis_client.cc#L135
`Connect(addresses[i], ports[i]`

The way this actually works, right now, the IP addresses are all always the same: the head uses its own IP address.
https://github.com/ray-project/ray/blob/master/python/ray/_private/services.py#L885
`shard_address = address(node_ip_address, redis_shard_port)`
But I assume that the *plan* here is for the head to eventually connect to other shards at other IP addresses, verify that everything's working, then publish those IP addresses for every joining node.

But I don't think that can possibly work in general.

In theoretical networks, every computer has exactly one IP address. In real networks, that often isn't true. The IP address by which the head node can reach a shard is not necessarily the IP address by which another node can reach the same shard.

On the other hand, this is a nice convenience feature to have. The shards might not be reachable by the same IP address the head used, but they might be, and if they are, that allows the user to skip providing the addresses on the secondary nodes.

Ultimately, I think the "right way" would be for every node to be able to specify `--redis-shard-addresses`.
https://docs.ray.io/en/master/configure.html#ports-configurations
But that would be a major project. This PR is a minor change that covers a lot of cases. (Indeed, at the moment it covers *all* the cases, since Ray does not currently permit placing the shards on a different host.)

This leaves the RedisShards addresses alone, but *if* we fail to connect to one of them, this tries a different IP address.

Which IP address? Well, there's only one IP address that was supplied by the user, and significantly, it's the IP address that (if we're reading the RedisShards from Redis) we've already successfully connected to.

At the moment, since https://github.com/ray-project/ray/blob/master/python/ray/_private/services.py#L885 is setting all the RedisShards to the head's IP address,
`shard_address = address(node_ip_address, redis_shard_port)`
it is in fact guaranteed that whatever IP address a node used to reach the head will also work to reach the shards.
Even later, if/when Ray enables placing shards on other hosts, this will cover the more common cases. If `<IP address>:6379` is successfully forwarded to where it belongs, then `<same IP address>:6380` will *probably* also be forwarded to where it belongs, without Ray needing to worry about exactly what is going on in the network.

(Obviously, if anyone is doing any forwarding, then the two networks are probably not exactly sitting on the same Infiniband (though they might be!). But latency between nodes, and minimizing communication between actors, is just the usual problem to deal with. If some particular actor is going to do a lot of communication and needs to be placed close to a data store, then Ray's delightful ability to specify custom resources can treat that proximity as a resource that actor requires.

Tying together machines on different subnetworks can be very useful if they have very different kinds of resources available --- lots of CPU power, lots of GPU power, or other "resources" like proximity to a database. And I think that'll actually be a pretty common situation, because having different resources correlates with being built by different people on different hardware for different original reasons. When those networks get grafted together --- keeping all their internal IP addresses and setting up some forwarding in between --- it would be very useful to be able to almost-seamlessly instantiate an actor and throw it over the fence to do its thing and call you back with an answer.)

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
Hmm, @dHannasch do you think we should just close 12026 and merge this instead?

**Comment by dHannasch** [comment]
> Hmm, @dHannasch do you think we should just close #12026 and merge this instead?

If you're happy with this.
(I was expecting that the only-emitting-a-warning version would be a rather easier sell than actually changing the address that Ray connects to.)

**Comment by dHannasch** [comment]
@richardliaw Now that PingPort is tested, does this look good?

**Comment by richardliaw** [comment]
Looks good to me, and nice that tests are passing. @rkooo567 can you merge (also pinged you offline).

@dHannasch can you join our slack? Would love to get in touch with you with a higher-bandwidth communication channel!

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
There are some unseen test failures. Can you merge the latest matser?

**Comment by dHannasch** [comment]
@rkooo567 Looks all that's failing on the latest are the Java and streaming tests (and the flaky test_actor on the msvc).

**Comment by rkooo567** [comment]
cc @ffbin Can you confirm if Java test failures are related to this PR?

File: src/ray/gcs/redis_client.cc [hunk 1]
```diff
@@ -132,6 +132,17 @@ Status RedisClient::Connect(std::vector<boost::asio::io_service *> io_services)
       boost::asio::io_service &io_service = *io_services[io_service_index];
       // Populate shard_contexts.
       shard_contexts_.push_back(std::make_shared<RedisContext>(io_service));
+      Status portReachable = shard_contexts_[i]->PingPort(addresses[i], ports[i]);
+      if (!portReachable.ok()) {
+        RAY_LOG(WARNING) << "Received Redis shard address " << addresses[i] << ":"
+                         << ports[i] << "from Redis head " << options_.server_ip_ << ":"
```
Comment (ffbin) [inline src/ray/gcs/redis_client.cc:138]:
```suggestion
                         << ports[i] << " from Redis head " << options_.server_ip_ << ":"
```

**Comment by ffbin** [comment]
> cc @ffbin Can you confirm if Java test failures are related to this PR?

Ok, i will take a look, thx.

**Comment by kfstorm** [comment]
@ffbin The Java and streaming tests are not related and have been fixed in master. @dHannasch Can you merge master again?

**Review by ffbin — APPROVED** [review]

**Comment by ffbin** [comment]
hi @dHannasch  I didn't encounter the failure of Windows CI job. I triggered CI job again.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**